File: java/Math/Sieve of Eratosthenes.java
Guarda en primes los números primos menores a MAX.
isPrime() retorna si p es o no un número primo.

static int MAX = 1000000;
static ArrayList<Integer> primes = new ArrayList<Integer>();
static boolean marked[] = new boolean[MAX/2];
	
static void sieve() {
	marked[0] = true;
	primes.add(2);
	for (long i = 3; i < MAX; i += 2) if (!marked[i/2]) {
		primes.add(i);
		for (long j = i*i; j < MAX; j += i*2) marked[j/2] = true;
	}
}

static boolean isPrime(int p) {
	if (p%2 != 0) return !sieve[p/2];
	return p == 2;
}

File: c++/Math/Sieve of Eratosthenes.cpp
Guarda en primes los numeros primos menores o iguales a MX. Para saber si p es un número primo, hacer: if (!marked[p]).

const int MX = 1e6;
bool marked[MX+1];
vector<int> primes;
// O(MX log(log(MX)))
void sieve() {
    marked[0] = marked[1] = true;
    for (int i = 2; i <= MX; i++) {
        if (marked[i]) continue;
        primes.pb(i);
        for (ll j = 1ll*i*i; j <= MX; j += i) marked[j] = true;
    }
}
